Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift

```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
                       "\(variationImageCount) variation images, \(variationAttributeCount) variation attributes")
         }
     }
+
+    func persistIncrementalCatalogData(_ catalog: POSCatalog, siteID: Int64) async throws {
+        DDLogInfo("💾 Persisting incremental catalog with \(catalog.products.count) products and \(catalog.variations.count) variations")
+
+        try await grdbManager.databaseConnection.write { db in
+            let site = PersistedSite(id: siteID)
+            try site.insert(db, onConflict: .ignore)
+
+            for product in catalog.productsToPersist {
+                try product.insert(db, onConflict: .replace)
+            }
+
+            let productIDs = catalog.products.map { $0.productID }.map { String($0) }.joined(separator: ",")
+
+            try PersistedProductImage
+                .filter(sql: "\(PersistedProductImage.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
+            for image in catalog.productImagesToPersist {
+                try image.insert(db, onConflict: .replace)
+            }
+
+            try PersistedProductAttribute
+                .filter(sql: "\(PersistedProductAttribute.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
+            for var attribute in catalog.productAttributesToPersist {
+                try attribute.insert(db)
+            }
+
+            for variation in catalog.variationsToPersist {
+                try variation.insert(db, onConflict: .replace)
+            }
+
+            let variationIDs = catalog.variations.map { $0.productVariationID }.map { String($0) }.joined(separator: ",")
+
+            try PersistedProductVariationImage
+                .filter(sql: "\(PersistedProductVariationImage.Columns.productVariationID.name) IN (\(variationIDs))")
+                .deleteAll(db)
+            for image in catalog.variationImagesToPersist {
+                try image.insert(db, onConflict: .replace)
```

Comment:
For the full sync, I used `onConflict: ignore`. We should be consistent – it doesn't matter which we go with for images or attributes (when we support global attributes in future.) Either first wins, or last wins.

For variations on an incremental sync, it does matter because we'll definitely want to replace... so perhaps we should use replace everywhere?

Reply:
Yea, incremental sync requires `relace` on conflict. For full sync, like you said, either works as we delete all the data at the beginning. It only matters if there are duplicate products/variations in the catalog (which is not impossible from the "Large fun testing" test site) with different values, but there's no way to know whether the first or last is correct in this edge case.

I replaced the `ignore` with `replace` on conflict strategy in full sync in
https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7162e20a2aeaaa912b06ded5c1f8912c9357a9a2.